New CSS reference: Keyframe selectors (#42573)
mdn/content: files/en-us/web/css/reference/selectors/keyframe/index.md

---
title: Keyframe selectors
slug: Web/CSS/Reference/Selectors/Keyframe
page-type: css-selector
browser-compat: css.at-rules.keyframes.selectors
sidebar: cssref
---

> CSS **keyframe selectors** identify specific points in an animation timeline where the keyframe's style should be applied. Defined in the [CSS animation module](/en-US/docs/Web/CSS/CSS_Animations), these selectors are used exclusively within the {{cssxref("@keyframes")}} at-rule.

Reviewer: ```suggestion
CSS **keyframe selectors** identify specific points in an animation timeline where keyframe styles should be applied. Defined in the [CSS animation module](/en-US/docs/Web/CSS/CSS_Animations), these selectors are used exclusively within the {{cssxref("@keyframes")}} at-rule.
```

...because it is often not a single keyframe.

> ## Syntax

Author: added

```css
/* Keywords */
from {
}
to {
}

/* <percentage> values */
0% {
}
50% {
}
100% {
}

/* Selector lists */
0%,
50%,
100% {
}
from,
to {
}
```

### Values

- `from`
  - : Represents the start of the animation sequence. It is equivalent to `0%`.
- `to`
  - : Represents the end of the animation sequence. It is equivalent to `100%`.
- {{cssxref("percentage")}}
  - : A percentage between `0%` and `100%`, inclusive, representing the total progression through the animation sequence.

## Description

> A `<keyframe-selector>` can be the `to` or `from` keyword, a percentage between `0%` and `100%`, inclusive, or a comma-separated list of these keywords and/or percentages. When a comma-separated list of `<keyframe-selector>` is used, the style block that follows applies to all the specified progression points.

Reviewer: ```suggestion
A `<keyframe-selector>` can be the `to` or `from` keyword, a percentage between `0%` and `100%`, inclusive, or a comma-separated list of these keywords and/or percentages. When a comma-separated `<keyframe-selector>` list is used, the style block that follows applies to all the specified progression points.
```

### Valid percentage values

Percentage values must include the percent sign (`%`). A unitless `0` is invalid. Values outside the range `0%` to `100%` (such as `-10%` or `110%`) are invalid and will cause the keyframe block to be ignored.

### Cascade, order, precedence, and importance

Property values set by animation `@keyframes` are more important than all normal styles, meaning no matter the {{cssxref("specificity")}}, an animated value will override values that do not have an {{cssxref("important", "!important")}} flag set. Only important property values and values currently being transitioned take precedence over animated property values.

All keyframe selectors have the same specificity. The order of keyframe selectors within the selector list does not matter. The source order only matters if a duplicate keyframe selector declares a different value for an already declared property on the same selector.

The `!important` flag is not valid within an `@keyframes` definition.

### Omitted start and end selectors

> If no `0%` (or `from`) or `100%` (or `to`) keyframe is specified, the browser will use the element's computed styles for those states, allowing the animation to transition smoothly from or to the element's non-animated property values. In other words, if a property is specified in a middle-of-the-timeline keyframe, without being set within a starting or ending keyframe selector block, the property will animate from it's original value to that value.

Reviewer: ```suggestion
If no `0%` (or `from`) or `100%` (or `to`) keyframe is specified, the browser will use the element's computed styles for those states, allowing the animation to transition smoothly from or to the element's non-animated property values. In other words, if a property is specified in a middle-of-the-timeline keyframe without being set within a starting or ending keyframe selector block, the property will animate from its original value to that value.
```

  from {
    transform: translatex(0);
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
  to {
    transform: translatex(50vw);
    opacity: 0;
  }
}
```

We apply our animation to the element using the {{cssxref("animation")}} shorthand property:

```css
div {
  animation: slide-and-fade 4s ease-in infinite;
}
```

#### Results

{{EmbedLiveSample("Basic usage","100%","200")}}

### Selector lists and initial values

Using the same HTML and CSS as in the previous example, this example demonstrates using comma-separated selectors to group selectors, applying the same styles at multiple points in an animation.

```html hidden
<div>I am animated</div>
```

#### CSS

We create a `pulse` animation that changes the size of our element. We also declare a `background-color` in only one selector block.

```css
div {
  animation: pulse 4s linear infinite;
}

@keyframes pulse {
  0%,
  75% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.2);
  }
  25%,
  100% {
    transform: scale(0.8);
    background-color: black;
  }
}
```

```css hidden
div {
  background-color: rebeccapurple;
  color: white;
  width: min-content;
  padding: 10px;
  font: 2rem sans-serif;
}
```

#### Result

> {{EmbedLiveSample("Selector lists and initial values","100%","125")}}

Reviewer: It is really hard to understand what is going on in this example. Can you simplify it a bit, use more obvious changes, or improve the description to explain it better?

Author: broke this into 2 examples

By using a selector list with multiple comma-separated keyframe selectors, the animation "pauses" from the `25%` keyframe until it reaches the `75%` keyframe.

> The `black` is only set in the `25%, 75%` keyframe, so the element animates from and to `rebeccapurple`, it's original `background-color`.

Reviewer: ```suggestion
The `black` background color is only set in the `25%, 100%` keyframe, so the element animates from and to `rebeccapurple`, its original `background-color`.
```

  }
  0%,
  50%,
  75%,
  100% {
    transform: translatex(0);
    opacity: 1;
    background-color: purple;
  }
}
```

### With named timeline ranges

Originally defined in the [CSS animation module](/en-US/docs/Web/CSS/Guides/Animations), the [CSS scroll-driven animations](/en-US/docs/Web/CSS/Guides/Scroll-driven_animations) module expanded the keyframe selector for view progress ranges. A {{cssxref("timeline-range-name")}} can precede the `<percentage>` component of the selector to attach keyframes to specific progress points within the named timeline range. The `<timeline-range-name>` represents the selected predefined named timeline range, and the `<percentage>` after it represents the percentage progress between the start and end of that named timeline range. This addition to the selector enables adding timeline range information directly in the `@keyframes` animation definition.

```css
@keyframes in-and-out {
  entry 0% {
    opacity: 0;
    transform: translateX(100%);
  }
  entry 100% {
    opacity: 1;
    transform: translateX(0);
  }
  exit 0% {
    opacity: 1;
    transform: translateX(0);
  }
  exit 100% {
    opacity: 0;
    transform: translateX(-100%);
  }
}
```

> If the element's animation timeline does not have a corresponding named timeline range, then any keyframes attached to points on that named timeline range are ignored. It is possible that these attachment points are outside the active interval of the animation. When this occurs, the automatic `from` (`0%`) and `to` (`100%`) keyframes are only generated for properties that don't have keyframes at or earlier than `0%` or at or after `100%`, respectively.

Reviewer: ```suggestion
If the element's animation timeline does not have a corresponding named view timeline range, then any keyframes attached to points on that named timeline range are ignored. These attachment points may be outside the active interval of the animation. When this occurs, the automatic `from` (`0%`) and `to` (`100%`) keyframes are only generated for properties that don't have keyframes at or earlier than `0%` or at or after `100%`, respectively.
```

```css
div {
  animation: slide-and-fade 4s ease-in infinite;
}
```

#### Results

{{EmbedLiveSample("Basic usage","100%","200")}}

### Selector lists

Using the same HTML and basic styling as in the previous example, this example demonstrates using comma-separated selectors to group selectors, applying the same styles at multiple points in an animation.

```html hidden
<div>I am animated</div>
```

#### CSS

We create a `pulse` animation that changes the size of our element.

```css
div {
  animation: pulse 4s linear infinite;
}

@keyframes pulse {
  0%,
  75% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.2);
  }
  25%,
  100% {
    transform: scale(0.8);
  }
}
```

```css hidden
div {
  background-color: rebeccapurple;
  color: white;
  width: min-content;
  padding: 10px;
  font: 2rem sans-serif;
}
```

#### Result

{{EmbedLiveSample("Selector lists and initial values","100%","125")}}

By using a selector list with multiple comma-separated keyframe selectors, the animation "pauses" from the `25%` keyframe until it reaches the `75%` keyframe.

### Omitting `to` and `from`

> This example demonstrates how, if starting or ending styles are omitted when the `to` or `from` keyframe selectors are not included in a `@keyframes` animation definition.

Reviewer: This doesn't make sense at present. I think you are missing a few words, but I'm not sure what you intended.

#### HTML

We include a few elements. We will be animating all of them.

```html
<div>I am animated</div>
<div>I am animated</div>
<div>I am animated</div>
```

#### CSS

We provide basic styles to our elements, giving each a different {{cssxref("outline-width")}} and {{cssxref("background-color")}}. These are the properties we will be animating.

```css
div {
  background-color: magenta;
  outline: 10px dashed black;
  color: white;
  width: min-content;
  padding: 10px;
  font: 2rem sans-serif;
  margin: 35px auto;

  animation: changes 5s linear infinite;
}
div:first-of-type {
  background-color: blue;
  outline-width: 0;
}
div:last-of-type {
  background-color: green;
  outline-width: 20px;
}
```

> We create an animation that sets an element's {{cssxref("border-width")}} and {{cssxref("background-color")}} at only the `30%` keyframe.

Reviewer: ```suggestion
We create an animation that sets an element's `background-color` and `outline-width` at `30%` and `40%` keyframes.
```
